qiskit/transpiler/passes/scheduling/dynamical_decoupling.py

```diff
@@ -0,0 +1,219 @@
+                    # absorb the inverse into the successor (from left in circuit)
+                    if succ.type == "op" and isinstance(succ.op, UGate):
+                        begin = int(slack / 2)
+                        new_dag.apply_operation_back(Delay(begin), [dag_qubit])
+                        new_dag.apply_operation_back(dd_sequence[0], [dag_qubit])
+                        new_dag.apply_operation_back(Delay(slack - begin), [dag_qubit])
+                        theta_r, phi_r, lam_r = succ.op.params
+                        succ.op.params = Optimize1qGates.compose_u3(
+                            theta_r, phi_r, lam_r, theta, phi, lam
+                        )
+                        new_dag.global_phase = _mod_2pi(new_dag.global_phase + phase)
+                    # absorb the inverse into the predecessor (from right in circuit)
+                    elif pred.type == "op" and isinstance(pred.op, UGate):
+                        begin = int(slack / 2)
+                        new_dag.apply_operation_back(Delay(begin), [dag_qubit])
+                        new_dag.apply_operation_back(dd_sequence[0], [dag_qubit])
+                        new_dag.apply_operation_back(Delay(slack - begin), [dag_qubit])
+                        theta_l, phi_l, lam_l = pred.op.params
+                        pred.op.params = Optimize1qGates.compose_u3(
+                            theta, phi, lam, theta_l, phi_l, lam_l
+                        )
+                        new_dag.global_phase = _mod_2pi(new_dag.global_phase + phase)
+                    # don't do anything if there's no single-qubit gate to absorb the inverse
+                    else:
+                        new_dag.apply_operation_back(nd.op, nd.qargs, nd.cargs)
+                else:
+                    # balanced spacing (d/2, d, d, ..., d, d, d/2)
+                    # careful here that we add up to the original delay duration
+                    num_pulses = len(self._dd_sequence)
+                    mid = int(slack / num_pulses)
+                    end = int(mid / 2)
+                    unused_slack = slack - 2 * end - (num_pulses - 1) * mid
+                    int((num_pulses - 1) / 2)
+                    delays = (
+                        [end]
+                        + [mid] * int((num_pulses - 1) / 2)
+                        + [mid + unused_slack]
+                        + [mid] * int((num_pulses - 1) / 2)
+                        + [end]
+                    )
+                    for gate, idle in itertools.zip_longest(self._dd_sequence, delays):
+                        new_dag.apply_operation_back(Delay(idle), [dag_qubit])
+                        if gate is not None:
+                            new_dag.apply_operation_back(gate, [dag_qubit])
+            else:
+                new_dag.apply_operation_back(nd.op, nd.qargs, nd.cargs)
+
+        new_dag.duration = dag.duration
```

Comment:
Should these be part of `dag._copy_circuit_metadata()`?

Reply:
Potentially, though I think duration as metadata is a bit dangerous because it can be overwritten easily. In reality duration should be calculated, similar to how we don't attach `depth` to the circuit. But for now I'll add it, requires a bit of though to improve.